Update mqtt.go

mandarj123 opened the pull request:
remove mandatory "-producer" to clientId

Signed-off-by: mandarj123 <mandar.jogalekar@capgemini.com>

# Description

While creating subscriber from Dapr for MQTT , the -producer is appended automatically.

This breaks the subscribers since remote mqtt broker cannot accept modified clientId
## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close

Fixes https://github.com/dapr/dapr/issues/5214
Fixes https://github.com/dapr/dapr/issues/5235

berndverst commented on pubsub/mqtt/mqtt.go:
```diff
@@ -89,7 +89,7 @@ func (m *mqttPubSub) Init(metadata pubsub.Metadata) error {
 
 	// mqtt broker allows only one connection at a given time from a clientID.
-	producerClientID := fmt.Sprintf("%s-producer", m.metadata.clientID)
+	producerClientID := m.metadata.clientID
```
This could be a breaking change. And why would you not also update the consumer with the same change?

If this PR is necessary (please explain in depth why) then instead I recommend introducing two new metadata options, `consumerClientID` and `producerClientID`. If those are set those values should be used, and if they are not set we should fall back to the current `clientID` metadata with the exact string we have today.

mandarj123 replied:
Hello, sure . I will have a look. I am not really a golang expert hence might take some time. But please consider this change as critical. since current version is not a working version if i want to use a remote MQTT publisher.

For eg: I am subscribing to a remote mqtt producer who provides me with certificates and clientID(eg. 12345)

When I provide this as consumerID to Dapr , dapr sends request to the producer with clientID:12345-producer

and this is breaking functionality. Producer will never have "-producer" in the clientID only to suit the subscriber.

Currently Dapr MQTT is built to only support both producers and subscribers in sync with each other and but it's not the case of the time. So this change is critical else we are not able to move ahead with out Dapr+MQTT implementation. Request you to consider this as critical.

ItalyPaleAle replied:
With MQTT, we need two different client IDs for producer and consumer because we maintain 2 connections to the service. Setting them to the same value wouldn't work.

However, I understand the issue @mandarj123 is reporting. One way we could fix that is by introducing 2 new metadata options: 

- `producerClientID`
- `consumerClientID`

If those are set, we will use their values for the producer/consumer client IDs. When they're unset (one of them or both), we will use the current default ("%s-producer" or "%s-consumer").

Thoughts?

mandarj123 replied:
I agree, and it would be great if someone can take this PR ahead and get this issue solved immediately.
My limited knowledge of golang means I could only do as many changes to remove the appending :)

ItalyPaleAle commented:
I updated the PR and added a new metadata option `producerID`. This changes the behavior so:

- If `consumerID` is set but **not** `producerID`, the current behavior remains (this is a stable component and we can't introduce backwards-incompatbile changes):
   - consumer is `{consumerID}-consumer`
   - producer is `{consumerID}-producer`
- if `producerID`  is set, then:
   - consumer is `{consumerID}`
   - producer is `{producerID}`
- Note that if `consumerID` is not set explicitly, the runtime sets it to the Dapr app ID automatically

Will update docs in a min

mandarj123 commented:
Thank you. Does this mean if I am consuming remote mqtt server . I will have to set producerId to a random string ? since I don’t publish messages but only subscribe using consumerId I will provide .





On 24 September 2022 at 12:20:23 AM IST, Alessandro (Ale) Segala ***@***.***> wrote:
This mail has been sent from an external source

I updated the PR and added a new metadata option producerID. This changes the behavior so:

  *   If consumerID is set but not producerID, the current behavior remains (this is a stable component and we can't introduce backwards-incompatbile changes):
     *   consumer is {consumerID}-consumer
     *   producer is {consumerID}-producer
  *   if producerID is set, then:
     *   consumer is {consumerID}
     *   producer is {producerID}
  *   Note that if consumerID is not set explicitly, the runtime sets it to the Dapr app ID automatically

Will update docs in a min

—
Reply to this email directly, view it on GitHub<https://github.com/dapr/components-contrib/pull/2104#issuecomment-1256555460>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AFO7BYFXFZMCDGYSPTOLG3LV7X3WZANCNFSM6AAAAAAQR7BSOE>.
You are receiving this because you were mentioned.Message ID: ***@***.***>
This message contains information that may be privileged or confidential and is the property of the Capgemini Group. It is intended only for the person to whom it is addressed. If you are not the intended recipient, you are not authorized to read, print, retain, copy, disseminate, distribute, or use this message or any part thereof. If you receive this message in error, please notify the sender immediately and delete all copies of this message.

ItalyPaleAle commented:
@mandarj123 yes. We can't do otherwise to preserve backwards compatibility, as MQTT is a stable component. (Unless @yaron2 / @berndverst say otherwise)

mandarj123 commented:
Thank you dapr team . Can you please help me understand when this will be completed and available for use ?





On 24 September 2022 at 4:10:39 AM IST, Bernd Verst ***@***.***> wrote:
This mail has been sent from an external source

@berndverst approved this pull request.

—
Reply to this email directly, view it on GitHub<https://github.com/dapr/components-contrib/pull/2104#pullrequestreview-1119096627>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AFO7BYATZZZHHM7GWDF2JOTV7YWWJANCNFSM6AAAAAAQR7BSOE>.
You are receiving this because you were assigned.Message ID: ***@***.***>
This message contains information that may be privileged or confidential and is the property of the Capgemini Group. It is intended only for the person to whom it is addressed. If you are not the intended recipient, you are not authorized to read, print, retain, copy, disseminate, distribute, or use this message or any part thereof. If you receive this message in error, please notify the sender immediately and delete all copies of this message.

berndverst commented:
This will be in our 1.9 release which should be in approximately 2 or 2.5 weeks.

mandarj123 commented:
Thanks bernd .


---
Sent from Workspace ONE Boxer<https://whatisworkspaceone.com/boxer>

On 24 September 2022 at 9:46:04 AM IST, Bernd Verst ***@***.***> wrote:
This mail has been sent from an external source

This will be in our 1.9 release which should be in approximately 2 or 2.5 weeks.

—
Reply to this email directly, view it on GitHub<https://github.com/dapr/components-contrib/pull/2104#issuecomment-1256851582>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AFO7BYALRUVNF64LOIULERLV7Z6ABANCNFSM6AAAAAAQR7BSOE>.
You are receiving this because you were assigned.Message ID: ***@***.***>
This message contains information that may be privileged or confidential and is the property of the Capgemini Group. It is intended only for the person to whom it is addressed. If you are not the intended recipient, you are not authorized to read, print, retain, copy, disseminate, distribute, or use this message or any part thereof. If you receive this message in error, please notify the sender immediately and delete all copies of this message.
